lib/oli/notifications.ex

```diff
@@ -0,0 +1,167 @@
+
+  @doc """
+  Returns the list of system_messages currently active.
+
+  ## Examples
+
+      iex> list_system_messages()
+      [%SystemMessage{}, ...]
+
+  """
+  def list_active_system_messages do
+    now = DateTime.utc_now()
+
+    Repo.all(
+      from(
+        sm in SystemMessage,
+        where: sm.active and ^now >= coalesce(sm.start, ^now) and ^now <= coalesce(sm.end, ^now),
+        order_by: [desc: :updated_at]
+      )
+    )
+  end
+
+  @doc """
+  Creates a system message.
+
+  ## Examples
+
+      iex> create_system_message(%{field: new_value})
+      {:ok, %SystemMessage{}}
+
+      iex> create_system_message(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def create_system_message(attrs \\ %{}) do
+    %SystemMessage{}
+    |> SystemMessage.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Gets a system message by id.
+
+  ## Examples
+
+      iex> get_system_message(1)
+      %SystemMessage{}
+      iex> get_system_message(123)
+      nil
+  """
+  def get_system_message(id), do: Repo.get(SystemMessage, id)
+
+  @doc """
+  Updates a system message and schedules jobs for broadcasting message updates.
+
+  ## Examples
+
+      iex> update_system_message(system_message, %{name: new_value})
+      {:ok, %SystemMessage{}}
+      iex> update_system_message(system_message, %{name: bad_value})
+      {:error, %Ecto.Changeset{}}
+  """
+  def update_system_message(%SystemMessage{} = system_message, attrs) do
+    system_message
+    |> SystemMessage.changeset(attrs)
+    |> Repo.update()
+    |> case do
+      {:ok, system_message} ->
+        schedule_message(system_message)
+        {:ok, system_message}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  @doc """
+  Deletes a system message.
+
+  ## Examples
+
+      iex> delete_system_message(system_message)
+      {:ok, %SystemMessage{}}
+
+      iex> delete_system_message(system_message)
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def delete_system_message(%SystemMessage{} = system_message), do: Repo.delete(system_message)
+
+  @doc """
+  Returns an `%Ecto.Changeset{}` for tracking system message changes.
+
+  ## Examples
+
+      iex> change_system_message(system_message)
+      %Ecto.Changeset{data: %SystemMessage{}}
+
+  """
+  def change_system_message(%SystemMessage{} = system_message, attrs \\ %{}) do
+    SystemMessage.changeset(system_message, attrs)
+  end
+
+  defp message_active_and_time_greater_than_now?(true, time) when not is_nil(time) do
+    now = DateTime.utc_now()
+
+    DateTime.compare(now, time) == :lt
+  end
+
+  defp message_active_and_time_greater_than_now?(_, _), do: false
+
+  defp schedule_message(
+         %SystemMessage{id: id, active: active, start: start_time, end: end_time} = system_message
+       ) do
+    remove_existing_message_jobs(id)
+
+    if message_active_and_time_greater_than_now?(active, start_time) do
```

Comment:
When I read the comments above I thought more on make the whole if a function and call it with the corresponding date and display value since that code is pretty much the same